PovioKit/Classes/Networking/RestClient/RestClient.swift

```diff
@@ -0,0 +1,160 @@
+                          _ result: ((Result<T, NetworkError>) -> Void)?) {
+    request(endpoint: endpoint, method: .post, parameters: parameters, headers: headers) {
+      switch $0 {
+      case .success(let response):
+        DispatchQueue.global(qos: .background).async {
+          do {
+            let decodedObject = try JSONDecoder.default.decode(T.self, from: response.data)
+            DispatchQueue.main.async { result?(.success(decodedObject)) }
+          } catch {
+            DispatchQueue.main.async { result?(.failure(.jsonDecode(error))) }
+          }
+        }
+      case .failure(let error):
+        DispatchQueue.main.async { result?(.failure(error)) }
+      }
+    }
+  }
+}
+
+// MARK: - Private Methods
+private extension RestClient {
+  func request(endpoint: EndpointProtocol,
+               method: HTTPMethod,
+               parameters: RestClientProtocol.Params? = nil,
+               headers: Headers? = nil,
+               _ result: RestClientProtocol.DataResult) {
+    updateToken { [weak self] token in
+      guard let strongSelf = self else { return }
+      Logger.debug("\(method.rawValue) (started): " + endpoint.path, params: ["url": endpoint.url])
+      let requestHeaders = self?.combineRequestHeaders(customHeaders: headers, token: token)
+      let encoding: ParameterEncoding = method == .get ? URLEncoding.default : JSONEncoding.default
+      strongSelf.manager
+        .request(endpoint.url, method: method, parameters: parameters, encoding: encoding, headers: requestHeaders)
+        .validate(statusCode: 200..<300)
+        .responseData { response in
+          let statusCode = response.response?.statusCode ?? 0
+          switch response.result {
+          case .success(let data):
+            Logger.debug("\(method.rawValue) (success): " + endpoint.path, params: ["status": statusCode, "url": endpoint.url])
+            let dataResponse = DataResponse(statusCode: response.response?.statusCode ?? 0, data: data)
+            strongSelf.retryCounter.clear(for: endpoint)
+            result?(.success(dataResponse))
+          case .failure(let error):
+            Logger.debug("\(method.rawValue) (failure): " + endpoint.path, params: ["status": statusCode, "url": endpoint.url])
+            switch strongSelf.retryCounter.shouldRetry(for: endpoint) {
+            case true:
+              let retryCount = strongSelf.retryCounter.retry(for: endpoint)
+              Logger.debug("\(method.rawValue) (retry): " + endpoint.path, params: ["retryAttempt": retryCount, "url": endpoint.url])
+              strongSelf.request(endpoint: endpoint, method: method, parameters: parameters, headers: headers, result)
+            case false:
+              strongSelf.retryCounter.clear(for: endpoint)
+              result?(.failure(.generic(error)))
+            }
+          }
+      }
+    }
+  }
+  
+  func updateToken(_ completion: ((_ token: String?) -> Void)?) {
```

Comment:
This method is firebase specific. We should throw it away for now.
And, we should think more about that to maybe define it at config stage 🤔